### PR TITLE
Get basic generators running in Limited API

### PR DESCRIPTION
**Pull request by da-woods**
This just seems to require removing the bits handling the f_back
of the current traceback.
    
It largely just brings it to the situation PyPy is in, so I imagine
some of the more complicated things based on this won't work.

### CODE REVIEW

**Comment by da-woods** [comment]
Draft because it's based on https://github.com/cython/cython/pull/5835

File: Cython/Utility/Coroutine.c [hunk 1]
```diff
@@ -1179,12 +1213,18 @@ static void __Pyx_Coroutine_dealloc(PyObject *self) {
         PyObject_ClearWeakRefs(self);
 
     if (gen->resume_label >= 0) {
+#if !CYTHON_USE_TP_FINALIZE
+        destructor del;
+#endif
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:1218]:
```suggestion
```

File: Cython/Utility/Coroutine.c [hunk 2]
```diff
@@ -1179,12 +1213,18 @@ static void __Pyx_Coroutine_dealloc(PyObject *self) {
...
         // Generator is paused or unstarted, so we need to close
         PyObject_GC_Track(self);
 #if CYTHON_USE_TP_FINALIZE
         if (unlikely(PyObject_CallFinalizerFromDealloc(self)))
 #else
-        Py_TYPE(gen)->tp_del(self);
+        del = __Pyx_PyObject_GetSlot(gen, tp_del, destructor);
+        if (del) {
+            del(self);
+        }
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:1227]:
```suggestion
        {
            destructor del = __Pyx_PyObject_GetSlot(gen, tp_del, destructor);
            if (del) del(self);
        }
```

File: Cython/Utility/Coroutine.c [hunk 3]
```diff
@@ -686,17 +711,25 @@ PyObject *__Pyx_Coroutine_SendEx(__pyx_CoroutineObject *self, PyObject *value, i
...
         #else
         // Generators always return to their most recent caller, not
         // necessarily their creator.
         PyObject *exc_tb;
         #if PY_VERSION_HEX >= 0x030B00a4 && !CYTHON_COMPILING_IN_CPYTHON
         // owned reference!
         exc_tb = PyException_GetTraceback(exc_state->exc_value);
+        #elif CYTHON_COMPILING_IN_LIMITED_API || CYTHON_COMPILING_IN_PYPY
+        PyObject *
         #elif PY_VERSION_HEX >= 0x030B00a4
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:733]: ?

File: Cython/Utility/Coroutine.c [hunk 4]
```diff
@@ -557,11 +567,26 @@ static int __Pyx_PyGen__FetchStopIterationValue(PyThreadState *$local_tstate_cna
...
+#endif
     Py_DECREF(ev);
+#if CYTHON_COMPILING_IN_LIMITED_API
+    if (!value) return -1;
+#endif
     *pvalue = value;
     return 0;
+
+#if !CYTHON_ASSUME_SAFE_MACROS
+  limited_api_failure:
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:584]:
```suggestion
#if CYTHON_COMPILING_IN_LIMITED_API || !CYTHON_ASSUME_SAFE_MACROS
  limited_api_failure:
```

**Review by scoder — COMMENTED** [review]
I would prefer positive guard conditions over negative ones (and I know that this PR predates such comments that I made elsewhere).

**Review by scoder — COMMENTED** [review]
I'm not completely done with the review, there's a lot to understand and look through here.

**Comment by da-woods** [comment]
@scoder Can you have a look at this when you have time?

File: Cython/Utility/Coroutine.c [hunk 5]
```diff
@@ -504,8 +506,13 @@ static int __Pyx_PyGen__FetchStopIterationValue(PyThreadState *$local_tstate_cna
             value = Py_None;
         }
         else if (likely(__Pyx_IS_TYPE(ev, (PyTypeObject*)PyExc_StopIteration))) {
+            #if CYTHON_COMPILING_IN_LIMITED_API
+            value = PyObject_GetAttrString(ev, "value");
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:510]: I think we can afford wasting a prepared Python string for this in non-limited API targets. Same further down.

File: Cython/Utility/Coroutine.c [hunk 6]
```diff
@@ -1381,15 +1423,24 @@ __Pyx_Coroutine_get_frame(__pyx_CoroutineObject *self, void *context)
...
+    __pyx_CoroutineObject *init_result;
     __pyx_CoroutineObject *gen = PyObject_GC_New(__pyx_CoroutineObject, type);
     if (unlikely(!gen))
         return NULL;
-    return __Pyx__Coroutine_NewInit(gen, body, code, closure, name, qualname, module_name);
+    init_result = __Pyx__Coroutine_NewInit(gen, body, code, closure, name, qualname, module_name);
+    if (unlikely(!init_result)) {
+        Py_DECREF(gen);
+    }
+    return init_result;
```
Comment (scoder) [inline Cython/Utility/Coroutine.c:1443]: This change doesn't seem necessary.

**Review by scoder — APPROVED** [review]
Looks good overall, just a few comments.

**Pull request merged**